gprofiler/heartbeat.py

```diff
@@ -0,0 +1,627 @@
+        self.current_thread: Optional[threading.Thread] = None
+        self.stop_event = threading.Event()
+        self.heartbeat_interval = 30  # seconds
+    
+    def start_heartbeat_loop(self):
+        """Start the main heartbeat loop"""
+        logger.info("Starting heartbeat loop...")
+        
+        while not self.stop_event.is_set():
+            try:
+                # Send heartbeat and check for commands
+                command_response = self.heartbeat_client.send_heartbeat()
+                
+                if command_response and command_response.get("profiling_command"):
+                    profiling_command = command_response["profiling_command"]
+                    command_id = command_response["command_id"]
+                    command_type = profiling_command.get("command_type", "start")
+                    
+                    logger.info(f"Received profiling command: {profiling_command}")
+                    
+                    # Check for idempotency - skip if command already executed
+                    if command_id in self.heartbeat_client.executed_command_ids:
+                        logger.info(f"Command ID {command_id} already executed, skipping...")
+
+                        # Wait for next heartbeat
+                        self.stop_event.wait(self.heartbeat_interval)
+
+                        continue
+                    
+                    logger.info(f"Received {command_type} command: {command_id}")
+                    
+                    # Mark command as executed for idempotency
+                    self.heartbeat_client.mark_command_executed(command_id)
+                    self.heartbeat_client.last_command_id = command_id
+                    
+                    if command_type == "stop":
+                        # Stop current profiler without starting a new one
+                        logger.info(f"RECEIVED STOP COMMAND for command ID: {command_id}")
+                        logger.info(f"STOP command details: {profiling_command}")
+                        self._stop_current_profiler()
+                        # TODO: important comment to make sure profiler has stopped successful to avoid leak 
+                        # Report completion for stop command
+                        self.heartbeat_client.send_command_completion(
+                            command_id=command_id,
+                            status="completed",
+                            execution_time=0,
+                            error_message=None,
+                            results_path=None
+                        )
+                    elif command_type == "start":
+                        # Stop current profiler if running, then start new one
+                        logger.info("Starting new profiler due to start command")
+                        # TODO: important comment to make sure profiler has stopped successful to avoid leak 
+                        self._stop_current_profiler()
```

Comment:
Please add the error handling for the cases stop and start are not processed properly.